lib/ghtorrent/etag_helper.rb

```diff
@@ -0,0 +1,154 @@
+require_relative 'paged_etag_match_error'
+class GHTorrent::EtagHelper
+  include GHTorrent::Settings
+
+  EMPTY_RESPONSE_ETAG = '5277b56db38c54d55de2903f41c9f6d6'
+
+  def initialize(command, url, use_etag = true)
+    @ght = command.ght
+    @url = url
+    @use_etag = use_etag
+  end
+
+  def request(media_type, paged)
+    result = verify_etag_and_get_response(media_type, paged) if first_page? && cacheable_endpoint?
+    result ||= @ght.do_request(@url, media_type)
+    store_etag_in_db(result, paged) if cacheable_endpoint? && !result.is_a?(String) && cacheable_page?(result.meta['link'])
+    result
+  rescue PagedEtagMatchError
+    # return nil for paged_api_request when response is not modified.
+  end
+
+  private
+
+  def verify_etag_and_get_response(media_type, paged)
+    etag_data, etag_response = etag_data_and_response(media_type)
+    return unless etag_response
+    log_etag_usage_and_raise_error(etag_data) if paged && not_modified?(etag_response)
+    # For single response & frontloaded api, this value will be 1.
+    # For modified backloaded api, this last page response is not useful.
+    return unless etag_data[:page_no] == 1
+    # Because we would be skipping store_etag_in_db for the String etag_data[:response].
+    increment_used_count if not_modified?(etag_response)
+    not_modified?(etag_response) ? etag_data[:response] : etag_response
+  end
+
+  def cacheable_endpoint?
+    return unless @use_etag
+    patterns = [%r{/user/(?:search|email)}, %r{/orgs/[^/]+/members},
+                %r{/users/[^/]+/orgs}, %r{/compare/.+\.\.\.},
+                %r{api.github.com/events},
+                %r{/commits/[^/]+$}, %r{/commits\?sha=}]
+    patterns.none? { |pattern| @url =~ pattern }
+  end
+
+  def etag_data_and_response(media_type)
+    etag_data = @ght.db[:etags].first(base_url: base_url) || empty_etag
```

Comment:
Thank you for raising this. We considered using MongoDB when designing this feature. At the time, we had two drivers. 

The first was that we were not sure the response storage in MongoDB would be persistent, that is; that the MongoDB data has some "expiration point", which would break sequential functionality that depends upon the response body being present. 

The second was that finding this response in MongoDB is far more complicated because the response is stored with the different API type, Repository, User, Commit, etc. That would drive explicit knowledge of *what* type of end point was being called in order to check whether there was an existing response or not.

We would like to note that the JSON response body is stored only for single object (not for collections) and is persisted only for a sub-set of single object such as repository and user. We expect that after the initial collection of these response bodies, the DB size will remain manageable.